specs/APIReview_CustomDownload.md

```diff
@@ -0,0 +1,283 @@
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    ## SecondFeatureName
+
+    Feature explanation text goes here, including why an app would use it, how it
+    replaces or supplements existing functionality.
+
+    ```c#
+    void SampleMethod()
+    {
+        var show = new AnExampleOf();
+        show.SomeMembers = AndWhyItMight(be, interesting)
+    }
+    ```
+
+    ```cpp
+    void SampleClass::SampleMethod()
+    {
+        winrt::com_ptr<ICoreWebView2> webview2 = ...
+    }
+    ```
+
+    As an example of this section, see the Examples section for the PasswordBox
+    control (https://docs.microsoft.com/windows/uwp/design/controls-and-patterns/password-box#examples).
+-->
+
+```cpp
+ScenarioCustomDownloadExperience::ScenarioCustomDownloadExperience(AppWindow* appWindow)
+    : m_appWindow(appWindow), m_webView(appWindow->GetWebView())
+{
+    //! [DownloadStarting]
+    // Register a handler for the `DownloadStarting` event.
+    // This example hides the default downloads UI and shows a dialog box instead.
+    // The dialog box displays the default save path and allows the user to specify a different path.
+    // Selecting `OK` will save the download to the chosen path.
+    // Selecting `Cancel` will cancel the download.
+
+    m_demoUri = L"https://demo.smartscreen.msft.net/";
+     CHECK_FAILURE(m_webView->add_DownloadStarting(
+         Callback<ICoreWebView2DownloadStartingEventHandler>(
+             [this](ICoreWebView2* sender, ICoreWebView2DownloadStartingEventArgs* args)
+                 -> HRESULT {
+
+                   // Hide the default downloads UI.
+                   CHECK_FAILURE(args->put_HideUI(TRUE));
+
+                   UINT64 totalBytes = 0;
+                   CHECK_FAILURE(args->get_TotalBytes(&totalBytes));
+
+                   wil::unique_cotaskmem_string url;
+                   CHECK_FAILURE(args->get_Url(&url));
+
+                   wil::unique_cotaskmem_string mimeType;
+                   CHECK_FAILURE(args->get_MimeType(&mimeType));
+
+                   wil::unique_cotaskmem_string contentDisposition;
+                   CHECK_FAILURE(args->get_ContentDisposition(&contentDisposition));
+
+                   wil::unique_cotaskmem_string savePath;
+                   CHECK_FAILURE(args->get_SavePath(&savePath));
+
+                  std::wstring prompt =
+                    std::wstring(L"Enter save path or select `OK` to use default path. ") +
+                    L"Select `Cancel` to cancel the download.";
+
+                  // Display the url, mime type, and total bytes of the download.
+                  std::wstring description = std::wstring(L"Url: ") + url.get() + L"\r\n" +
+                    L"Mime type: " + mimeType.get() + L"\r\n" +
+                    L"Total bytes: " + std::to_wstring(totalBytes) + L"\r\n";
+
+                   TextInputDialog dialog(
```

Comment:
Seems like we need to support deferrals for this event. The end dev may need to do UI to interact with the user (like in this example) that may require running code async (unlike this example) in which case we should provide a deferral for the end dev to take until they're done doing async work.

Reply:
Thanks for the feedback! I added deferral to the IDL, but do the examples also need to change to demonstrate what you described?

Reply:
Yes that would be good. Previous PRs feedback was not to have sync dialogs like this during event handlers. Instead we'd need to have the event handler start something async to show the dialog and use the deferral. Hopefully we already have an example in our sample app of using a deferral like this.